File: src/db/token.rs
use std::collections::HashSet;

use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use rand::Rng;
use sha2::{Digest, Sha256};
use sqlx::SqlitePool;

use crate::error::AppError;
use crate::models::token::{
    utc_now_timestamp, ApiTokenCreatedOut, ApiTokenIn, ApiTokenOut, ApiTokenRow, ApiTokenUpdateIn,
};

const TOKEN_PREVIEW_CHARS: usize = 8;

/// Every column of a token row, in the order `ApiTokenRow` declares them.
const TOKEN_COLUMNS: &str =
    "id, name, description, token_preview, enabled, expires_at, created_at, updated_at";

// ── Helpers ────────────────────────────────────────────────────────────────────────────

pub fn generate_api_token() -> String {
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill(&mut bytes);
    let encoded = URL_SAFE_NO_PAD.encode(bytes);
    format!("wildtoken_{encoded}")
}

pub fn token_digest(token: &str) -> String {
    format!("{:x}", Sha256::digest(token.as_bytes()))
}

pub fn token_preview(token: &str) -> String {
    let char_count = token.chars().count();
    let visible_chars = if char_count > TOKEN_PREVIEW_CHARS {
        TOKEN_PREVIEW_CHARS
    } else {
        char_count / 2
    };
    let prefix: String = token.chars().take(visible_chars).collect();
    format!("{prefix}…")
}

fn token_out(row: ApiTokenRow) -> ApiTokenOut {
    let ApiTokenRow {
        id,
        name,
        description,
        token_preview,
        enabled,
        expires_at,
        created_at,
        updated_at,
    } = row;
    ApiTokenOut {
        id,
        name,
        description,
        token_preview,
        enabled: enabled == 1,
        expires_at,
        created_at,
        updated_at,
    }
}

/// Reject an expiry that has already passed.
///
/// Compared as text rather than as parsed instants: `expires_at` and SQLite's
/// `datetime('now')` share one fixed-width UTC shape, so this reaches the same
/// verdict as the authentication SQL in `middleware::auth`. Parsing here
/// instead would open a window where the console accepts an expiry the proxy
/// already treats as dead.
fn reject_past_expiry(expires_at: Option<&str>) -> Result<(), AppError> {
    match expires_at {
        Some(expiry) if expiry <= utc_now_timestamp().as_str() => Err(AppError::BadRequest(
            "token expiry must be in the future; disable or delete the token to revoke it now"
                .into(),
        )),
        _ => Ok(()),
    }
}

/// Upgrade plaintext token rows in one transaction.
///
/// The legacy `token` column is retained because its NOT NULL/UNIQUE constraints
/// are part of deployed databases. Its values are overwritten with the same
/// SHA-256 digest stored in `token_hash`, so no plaintext survives the commit.
pub async fn migrate_legacy_token_storage(pool: &SqlitePool) -> Result<(), sqlx::Error> {
    let mut tx = pool.begin().await?;
    let columns: Vec<String> =
        sqlx::query_scalar("SELECT name FROM pragma_table_info('api_tokens') ORDER BY cid")
            .fetch_all(&mut *tx)
            .await?;
    let has_legacy_token = columns.iter().any(|column| column == "token");
    let has_token_hash = columns.iter().any(|column| column == "token_hash");
    let has_token_preview = columns.iter().any(|column| column == "token_preview");

    if !has_legacy_token {
        return Err(sqlx::Error::Protocol(
            "api_tokens is missing its compatibility token column".into(),
        ));
    }
    if !has_token_hash {
        sqlx::query("ALTER TABLE api_tokens ADD COLUMN token_hash TEXT")
            .execute(&mut *tx)
            .await?;
    }
    if !has_token_preview {
        sqlx::query("ALTER TABLE api_tokens ADD COLUMN token_preview TEXT NOT NULL DEFAULT ''")
            .execute(&mut *tx)
            .await?;
    }

    if !has_token_hash {
        let rows: Vec<(i64, String)> =
            sqlx::query_as("SELECT id, token FROM api_tokens ORDER BY id")
                .fetch_all(&mut *tx)
                .await?;
        let existing_tokens: HashSet<&str> = rows.iter().map(|(_, token)| token.as_str()).collect();

        let marker_prefix = loop {
            let mut bytes = [0_u8; 16];
            rand::thread_rng().fill(&mut bytes);
            let candidate = format!(
                "__wildtoken_token_migration_{}__",
                URL_SAFE_NO_PAD.encode(bytes)
            );
            if existing_tokens
                .iter()
                .all(|token| !token.starts_with(&candidate))
            {
                break candidate;
            }
        };

        for (id, plaintext) in rows {
            sqlx::query(
                "UPDATE api_tokens SET token = ?, token_hash = ?, token_preview = ? WHERE id = ?",
            )
            .bind(format!("{marker_prefix}{id}"))
            .bind(token_digest(&plaintext))
            .bind(token_preview(&plaintext))
            .bind(id)
            .execute(&mut *tx)
            .await?;
        }
    }

    let missing_hashes: i64 =
        sqlx::query_scalar("SELECT COUNT(*) FROM api_tokens WHERE token_hash IS NULL")
            .fetch_one(&mut *tx)
            .await?;
    if missing_hashes != 0 {
        return Err(sqlx::Error::Protocol(
            "api_tokens contains rows without a token digest".into(),
        ));
    }

    // Clear compatibility-column plaintext after every startup. This is also a
    // repair guard for databases whose legacy column was modified out of band.
    sqlx::query("UPDATE api_tokens SET token = token_hash WHERE token <> token_hash")
        .execute(&mut *tx)
        .await?;
    sqlx::query(
        "CREATE UNIQUE INDEX IF NOT EXISTS idx_api_tokens_token_hash ON api_tokens(token_hash)",
    )
    .execute(&mut *tx)
    .await?;

    tx.commit().await
}

// ── Public functions ─────────────────────────────────────────────────────────────────

pub async fn list_tokens(pool: &SqlitePool) -> Result<Vec<ApiTokenOut>, AppError> {
    let rows: Vec<ApiTokenRow> = sqlx::query_as(&format!(
        "SELECT {TOKEN_COLUMNS} FROM api_tokens ORDER BY id ASC"
    ))
    .fetch_all(pool)
    .await?;

    Ok(rows.into_iter().map(token_out).collect())
}

pub async fn get_token(pool: &SqlitePool, id: i64) -> Result<Option<ApiTokenOut>, AppError> {
    let row: Option<ApiTokenRow> = sqlx::query_as(&format!(
        "SELECT {TOKEN_COLUMNS} FROM api_tokens WHERE id = ?"
    ))
    .bind(id)
    .fetch_optional(pool)
    .await?;

    Ok(row.map(token_out))
}

pub async fn create_token(
    pool: &SqlitePool,
    input: &ApiTokenIn,
) -> Result<ApiTokenCreatedOut, AppError> {
    input
        .validate()
        .map_err(|message| AppError::BadRequest(message.into()))?;
    let expires_at = input
        .normalized_expires_at()
        .map_err(|message| AppError::BadRequest(message.into()))?;
    reject_past_expiry(expires_at.as_deref())?;

    let token_value = input.token.clone().unwrap_or_else(generate_api_token);
    let token_hash = token_digest(&token_value);
    let token_preview = token_preview(&token_value);
    let enabled = i64::from(input.enabled);

    let result = sqlx::query(
        r#"INSERT INTO api_tokens
        (name, description, token, token_hash, token_preview, enabled, expires_at, created_at, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, datetime('now'), datetime('now'))"#,
    )
    .bind(input.name.trim())
    .bind(input.description.trim())
    .bind(&token_hash)
    .bind(&token_hash)
    .bind(&token_preview)
    .bind(enabled)
    .bind(expires_at.as_deref())
    .execute(pool)
    .await?;

    let last_id = result.last_insert_rowid();
    let row: ApiTokenRow = sqlx::query_as(&format!(
        "SELECT {TOKEN_COLUMNS} FROM api_tokens WHERE id = ?"
    ))
    .bind(last_id)
    .fetch_one(pool)
    .await?;

    Ok(ApiTokenCreatedOut {
        id: row.id,
        name: row.name,
        description: row.description,
        token: token_value,
        token_preview: row.token_preview,
        enabled: row.enabled == 1,
        expires_at: row.expires_at,
        created_at: row.created_at,
        updated_at: row.updated_at,
    })
}

pub async fn update_token(
    pool: &SqlitePool,
    id: i64,
    input: &ApiTokenUpdateIn,
) -> Result<ApiTokenOut, AppError> {
    input
        .validate()
        .map_err(|message| AppError::BadRequest(message.into()))?;
    let expires_at = input
        .normalized_expires_at()
        .map_err(|message| AppError::BadRequest(message.into()))?;

    let existing = get_token(pool, id)
        .await?
        .ok_or_else(|| AppError::NotFound("token not found".into()))?;

    // Only a changed expiry has to be in the future. Editing a token that has
    // already lapsed sends its own past expiry back unchanged, and rejecting
    // that would make renaming an expired token impossible without renewing it.
    if expires_at != existing.expires_at {
        reject_past_expiry(expires_at.as_deref())?;
    }

    sqlx::query(
        "UPDATE api_tokens SET name = ?, description = ?, expires_at = ?, updated_at = datetime('now') WHERE id = ?",
    )
    .bind(input.name.trim())
    .bind(input.description.trim())
    .bind(expires_at.as_deref())
    .bind(id)
    .execute(pool)
    .await?;

    get_token(pool, id)
        .await?
        .ok_or_else(|| AppError::NotFound("token not found".into()))
}

pub async fn set_token_enabled(
    pool: &SqlitePool,
    id: i64,
    enabled: bool,
) -> Result<ApiTokenOut, AppError> {
    let val = i64::from(enabled);

    sqlx::query("UPDATE api_tokens SET enabled = ?, updated_at = datetime('now') WHERE id = ?")
        .bind(val)
        .bind(id)
        .execute(pool)
        .await?;

    get_token(pool, id)
        .await?
        .ok_or_else(|| AppError::NotFound("token not found".into()))
}

pub async fn delete_token(pool: &SqlitePool, id: i64) -> Result<bool, AppError> {
    let result = sqlx::query("DELETE FROM api_tokens WHERE id = ?")
        .bind(id)
        .execute(pool)
        .await?;

    Ok(result.rows_affected() > 0)
}

#[cfg(test)]
mod tests {
    use sqlx::sqlite::SqlitePoolOptions;

    use super::{create_token, get_token, token_digest, token_preview, update_token};
    use crate::{
        error::AppError,
        models::token::{ApiTokenIn, ApiTokenUpdateIn},
        state::init_db,
    };

    async fn memory_pool() -> sqlx::SqlitePool {
        let pool = SqlitePoolOptions::new()
            .max_connections(1)
            .connect("sqlite::memory:")
            .await
            .unwrap();
        init_db(&pool).await.unwrap();
        pool
    }

    fn create_input(name: &str, expires_at: Option<&str>) -> ApiTokenIn {
        ApiTokenIn {
            name: name.into(),
            description: String::new(),
            token: None,
            enabled: true,
            expires_at: expires_at.map(str::to_owned),
        }
    }

    #[tokio::test]
    async fn creation_honors_enabled_and_persists_only_the_digest() {
        let pool = memory_pool().await;
        let plaintext = "custom-token-value-1234";

        let created = create_token(
            &pool,
            &ApiTokenIn {
                name: " disabled client ".into(),
                description: " disabled at creation ".into(),
                token: Some(plaintext.into()),
                enabled: false,
                expires_at: None,
            },
        )
        .await
        .unwrap();

        assert_eq!(created.token, plaintext);
        assert!(!created.enabled);
        assert_eq!(created.name, "disabled client");
        let stored: (String, String, String, i64) = sqlx::query_as(
            "SELECT token, token_hash, token_preview, enabled FROM api_tokens WHERE id = ?",
        )
        .bind(created.id)
        .fetch_one(&pool)
        .await
        .unwrap();
        assert_eq!(stored.0, token_digest(plaintext));
        assert_eq!(stored.1, token_digest(plaintext));
        assert_ne!(stored.2, plaintext);
        assert_eq!(stored.3, 0);
    }

    #[tokio::test]
    async fn creation_stores_a_normalized_expiry_and_refuses_a_past_one() {
        let pool = memory_pool().await;

        let never = create_token(&pool, &create_input("never", None))
            .await
            .unwrap();
        assert_eq!(never.expires_at, None);

        // An offset form is accepted and lands in the stored UTC shape.
        let dated = create_token(
            &pool,
            &create_input("dated", Some("2099-01-01T08:00:00+08:00")),
        )
        .await
        .unwrap();
        assert_eq!(dated.expires_at.as_deref(), Some("2099-01-01 00:00:00"));

        let past = create_token(&pool, &create_input("past", Some("2000-01-01 00:00:00"))).await;
        assert!(matches!(past, Err(AppError::BadRequest(_))));
    }

    #[tokio::test]
    async fn updating_can_renew_or_clear_an_expiry_and_leaves_a_lapsed_one_editable() {
        let pool = memory_pool().await;
        let created = create_token(&pool, &create_input("client", Some("2099-01-01 00:00:00")))
            .await
            .unwrap();

        // Backdate the row the way time would have.
        sqlx::query("UPDATE api_tokens SET expires_at = '2000-01-01 00:00:00' WHERE id = ?")
            .bind(created.id)
            .execute(&pool)
            .await
            .unwrap();

        // Renaming an expired token resends its own past expiry unchanged. That
        // must not be rejected, or the row could never be edited again.
        let renamed = update_token(
            &pool,
            created.id,
            &ApiTokenUpdateIn {
                name: "renamed".into(),
                description: String::new(),
                expires_at: Some("2000-01-01 00:00:00".into()),
            },
        )
        .await
        .unwrap();
        assert_eq!(renamed.name, "renamed");
        assert_eq!(renamed.expires_at.as_deref(), Some("2000-01-01 00:00:00"));

        // Moving it to a different past instant is a real change, so it is not.
        let backdated = update_token(
            &pool,
            created.id,
            &ApiTokenUpdateIn {
                name: "renamed".into(),
                description: String::new(),
                expires_at: Some("2001-01-01 00:00:00".into()),
            },
        )
        .await;
        assert!(matches!(backdated, Err(AppError::BadRequest(_))));

        let renewed = update_token(
            &pool,
            created.id,
            &ApiTokenUpdateIn {
                name: "renamed".into(),
                description: String::new(),
                expires_at: Some("2099-06-01 00:00:00".into()),
            },
        )
        .await
        .unwrap();
        assert_eq!(renewed.expires_at.as_deref(), Some("2099-06-01 00:00:00"));

        // An absent expiry clears it — the update endpoint is a full replacement.
        update_token(
            &pool,
            created.id,
            &ApiTokenUpdateIn {
                name: "renamed".into(),
                description: String::new(),
                expires_at: None,
            },
        )
        .await
        .unwrap();
        assert_eq!(
            get_token(&pool, created.id)
                .await
                .unwrap()
                .unwrap()
                .expires_at,
            None
        );
    }

    #[test]
    fn preview_never_contains_a_short_token_in_full() {
        assert_eq!(token_preview(""), "…");
        assert_eq!(token_preview("x"), "…");
        assert_eq!(token_preview("short"), "sh…");
        assert_eq!(token_preview("long-enough-token"), "long-eno…");
    }
}

File: internal/db/token.go
package db

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"database/sql"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"fmt"

	"github.com/liguangsheng/wildtoken/internal/apperr"
	"github.com/liguangsheng/wildtoken/internal/models"
)

const tokenPreviewChars = 8

// tokenColumns lists every column of a token row, in the order APITokenRow
// declares them.
const tokenColumns = `t.id, t.name, t.description, t.token_preview, t.enabled,
    t.expires_at, t.created_at, t.updated_at,
    COALESCE(t.group_id, 1),
    COALESCE((SELECT g.name FROM groups g WHERE g.id = t.group_id), 'default')`

// tokenFrom is the FROM clause matching tokenColumns.
const tokenFrom = " FROM api_tokens AS t"

// GenerateAPIToken mints a fresh downstream token.
func GenerateAPIToken() (string, error) {
	bytes := make([]byte, 32)
	if _, err := rand.Read(bytes); err != nil {
		return "", apperr.Internal("could not generate an API token")
	}
	return "wildtoken_" + base64.RawURLEncoding.EncodeToString(bytes), nil
}

// TokenDigest is the only form of a token that is ever persisted.
func TokenDigest(token string) string {
	sum := sha256.Sum256([]byte(token))
	return hex.EncodeToString(sum[:])
}

// TokenPreview renders the console hint, never revealing a short token in full.
func TokenPreview(token string) string {
	runes := []rune(token)
	visible := len(runes) / 2
	if len(runes) > tokenPreviewChars {
		visible = tokenPreviewChars
	}
	return string(runes[:visible]) + "…"
}

func scanTokenRow(row interface{ Scan(...any) error }) (models.APITokenRow, string, error) {
	var token models.APITokenRow
	var expiresAt sql.NullString
	var groupName string
	err := row.Scan(&token.ID, &token.Name, &token.Description, &token.TokenPreview,
		&token.Enabled, &expiresAt, &token.CreatedAt, &token.UpdatedAt,
		&token.GroupID, &groupName)
	if err != nil {
		return token, "", err
	}
	if expiresAt.Valid {
		token.ExpiresAt = &expiresAt.String
	}
	return token, groupName, nil
}

func tokenOut(row models.APITokenRow, groupName string) models.APITokenOut {
	return models.APITokenOut{
		ID:           row.ID,
		Name:         row.Name,
		Description:  row.Description,
		TokenPreview: row.TokenPreview,
		Enabled:      row.Enabled == 1,
		ExpiresAt:    row.ExpiresAt,
		CreatedAt:    row.CreatedAt,
		UpdatedAt:    row.UpdatedAt,
		GroupID:      row.GroupID,
		GroupName:    groupName,
	}
}

// rejectPastExpiry refuses an expiry that has already passed.
//
// Compared as text rather than as parsed instants: expires_at and SQLite's
// datetime('now') share one fixed-width UTC shape, so this reaches the same
// verdict as the authentication SQL in the auth middleware. Parsing here instead
// would open a window where the console accepts an expiry the proxy already
// treats as dead.
func rejectPastExpiry(expiresAt *string) error {
	if expiresAt == nil {
		return nil
	}
	if *expiresAt <= models.UTCNowTimestamp() {
		return apperr.BadRequest(
			"token expiry must be in the future; disable or delete the token to revoke it now")
	}
	return nil
}

// MigrateLegacyTokenStorage upgrades plaintext token rows in one transaction.
//
// The legacy `token` column is retained because its NOT NULL/UNIQUE constraints
// are part of deployed databases. Its values are overwritten with the same
// SHA-256 digest stored in `token_hash`, so no plaintext survives the commit.
func MigrateLegacyTokenStorage(ctx context.Context, db *sql.DB) error {
	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()

	columns, err := tableColumns(ctx, tx, "api_tokens")
	if err != nil {
		return err
	}
	hasLegacyToken := columns["token"]
	hasTokenHash := columns["token_hash"]
	hasTokenPreview := columns["token_preview"]

	if !hasLegacyToken {
		return errors.New("api_tokens is missing its compatibility token column")
	}
	if !hasTokenHash {
		if _, err := tx.ExecContext(ctx, "ALTER TABLE api_tokens ADD COLUMN token_hash TEXT"); err != nil {
			return err
		}
	}
	if !hasTokenPreview {
		if _, err := tx.ExecContext(ctx,
			"ALTER TABLE api_tokens ADD COLUMN token_preview TEXT NOT NULL DEFAULT ''"); err != nil {
			return err
		}
	}

	if !hasTokenHash {
		if err := hashLegacyPlaintextTokens(ctx, tx); err != nil {
			return err
		}
	}

	var missingHashes int64
	if err := tx.QueryRowContext(ctx,
		"SELECT COUNT(*) FROM api_tokens WHERE token_hash IS NULL").Scan(&missingHashes); err != nil {
		return err
	}
	if missingHashes != 0 {
		return errors.New("api_tokens contains rows without a token digest")
	}

	// Clear compatibility-column plaintext after every startup. This is also a
	// repair guard for databases whose legacy column was modified out of band.
	if _, err := tx.ExecContext(ctx,
		"UPDATE api_tokens SET token = token_hash WHERE token <> token_hash"); err != nil {
		return err
	}
	if _, err := tx.ExecContext(ctx,
		"CREATE UNIQUE INDEX IF NOT EXISTS idx_api_tokens_token_hash ON api_tokens(token_hash)"); err != nil {
		return err
	}

	return tx.Commit()
}

func tableColumns(ctx context.Context, tx *sql.Tx, table string) (map[string]bool, error) {
	rows, err := tx.QueryContext(ctx,
		fmt.Sprintf("SELECT name FROM pragma_table_info('%s') ORDER BY cid", table))
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	columns := map[string]bool{}
	for rows.Next() {
		var name string
		if err := rows.Scan(&name); err != nil {
			return nil, err
		}
		columns[name] = true
	}
	return columns, rows.Err()
}

// hashLegacyPlaintextTokens replaces every plaintext value with its digest,
// using a collision-free marker for the legacy UNIQUE column.
func hashLegacyPlaintextTokens(ctx context.Context, tx *sql.Tx) error {
	rows, err := tx.QueryContext(ctx, "SELECT id, token FROM api_tokens ORDER BY id")
	if err != nil {
		return err
	}
	type legacyToken struct {
		id        int64
		plaintext string
	}
	var legacy []legacyToken
	for rows.Next() {
		var entry legacyToken
		if err := rows.Scan(&entry.id, &entry.plaintext); err != nil {
			rows.Close()
			return err
		}
		legacy = append(legacy, entry)
	}
	rows.Close()
	if err := rows.Err(); err != nil {
		return err
	}

	markerPrefix, err := uniqueMarkerPrefix(legacy)
	if err != nil {
		return err
	}

	for _, entry := range legacy {
		_, err := tx.ExecContext(ctx,
			"UPDATE api_tokens SET token = ?, token_hash = ?, token_preview = ? WHERE id = ?",
			fmt.Sprintf("%s%d", markerPrefix, entry.id),
			TokenDigest(entry.plaintext), TokenPreview(entry.plaintext), entry.id)
		if err != nil {
			return err
		}
	}
	return nil
}

func uniqueMarkerPrefix[T any](rows []T) (string, error) {
	_ = rows
	bytes := make([]byte, 16)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return fmt.Sprintf("__wildtoken_token_migration_%s__",
		base64.RawURLEncoding.EncodeToString(bytes)), nil
}

func ListTokens(ctx context.Context, db *sql.DB) ([]models.APITokenOut, error) {
	rows, err := db.QueryContext(ctx,
		"SELECT "+tokenColumns+tokenFrom+" ORDER BY t.id ASC")
	if err != nil {
		return nil, apperr.Database(err)
	}
	defer rows.Close()

	tokens := []models.APITokenOut{}
	for rows.Next() {
		row, groupName, err := scanTokenRow(rows)
		if err != nil {
			return nil, apperr.Database(err)
		}
		tokens = append(tokens, tokenOut(row, groupName))
	}
	if err := rows.Err(); err != nil {
		return nil, apperr.Database(err)
	}
	return tokens, nil
}

// GetToken returns ok=false when no row carries the id.
func GetToken(ctx context.Context, db *sql.DB, id int64) (models.APITokenOut, bool, error) {
	row := db.QueryRowContext(ctx, "SELECT "+tokenColumns+tokenFrom+" WHERE t.id = ?", id)
	entry, groupName, err := scanTokenRow(row)
	if errors.Is(err, sql.ErrNoRows) {
		return models.APITokenOut{}, false, nil
	}
	if err != nil {
		return models.APITokenOut{}, false, apperr.Database(err)
	}
	return tokenOut(entry, groupName), true, nil
}

func CreateToken(ctx context.Context, db *sql.DB, input *models.APITokenIn) (models.APITokenCreatedOut, error) {
	if err := input.Validate(); err != nil {
		return models.APITokenCreatedOut{}, apperr.BadRequest(err.Error())
	}
	expiresAt, err := input.NormalizedExpiresAt()
	if err != nil {
		return models.APITokenCreatedOut{}, apperr.BadRequest(err.Error())
	}
	if err := rejectPastExpiry(expiresAt); err != nil {
		return models.APITokenCreatedOut{}, err
	}

	tokenValue := ""
	if input.Token != nil {
		tokenValue = *input.Token
	} else {
		if tokenValue, err = GenerateAPIToken(); err != nil {
			return models.APITokenCreatedOut{}, err
		}
	}
	digest := TokenDigest(tokenValue)
	preview := TokenPreview(tokenValue)

	groupID, err := resolveTokenGroup(ctx, db, input.GroupID)
	if err != nil {
		return models.APITokenCreatedOut{}, err
	}

	result, err := db.ExecContext(ctx, `INSERT INTO api_tokens
        (name, description, token, token_hash, token_preview, enabled, expires_at, group_id, created_at, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, datetime('now'), datetime('now'))`,
		trimSpace(input.Name), trimSpace(input.Description), digest, digest, preview,
		boolToInt64(input.Enabled), expiresAt, groupID)
	if err != nil {
		return models.APITokenCreatedOut{}, apperr.Database(err)
	}
	id, err := result.LastInsertId()
	if err != nil {
		return models.APITokenCreatedOut{}, apperr.Database(err)
	}

	created, ok, err := GetToken(ctx, db, id)
	if err != nil {
		return models.APITokenCreatedOut{}, err
	}
	if !ok {
		return models.APITokenCreatedOut{}, apperr.Internal("token was not persisted")
	}

	return models.APITokenCreatedOut{
		ID:           created.ID,
		Name:         created.Name,
		Description:  created.Description,
		Token:        tokenValue,
		TokenPreview: created.TokenPreview,
		Enabled:      created.Enabled,
		ExpiresAt:    created.ExpiresAt,
		CreatedAt:    created.CreatedAt,
		UpdatedAt:    created.UpdatedAt,
		GroupID:      created.GroupID,
		GroupName:    created.GroupName,
	}, nil
}

// resolveTokenGroup validates a requested group, defaulting when absent.
//
// A token must always name a group that exists: with no group it would reach no
// channel, which reads as a routing bug rather than a configuration choice.
func resolveTokenGroup(ctx context.Context, db *sql.DB, requested *int64) (int64, error) {
	groupID := models.DefaultGroupID
	if requested != nil && *requested > 0 {
		groupID = *requested
	}
	exists, err := GroupExists(ctx, db, groupID)
	if err != nil {
		return 0, err
	}
	if !exists {
		return 0, apperr.BadRequest("group does not exist")
	}
	return groupID, nil
}

func UpdateToken(ctx context.Context, db *sql.DB, id int64, input *models.APITokenUpdateIn) (models.APITokenOut, error) {
	if err := input.Validate(); err != nil {
		return models.APITokenOut{}, apperr.BadRequest(err.Error())
	}
	expiresAt, err := input.NormalizedExpiresAt()
	if err != nil {
		return models.APITokenOut{}, apperr.BadRequest(err.Error())
	}

	existing, ok, err := GetToken(ctx, db, id)
	if err != nil {
		return models.APITokenOut{}, err
	}
	if !ok {
		return models.APITokenOut{}, apperr.NotFound("token not found")
	}

	// Only a changed expiry has to be in the future. Editing a token that has
	// already lapsed sends its own past expiry back unchanged, and rejecting
	// that would make renaming an expired token impossible without renewing it.
	if !equalOptionalString(expiresAt, existing.ExpiresAt) {
		if err := rejectPastExpiry(expiresAt); err != nil {
			return models.APITokenOut{}, err
		}
	}

	groupID, err := resolveTokenGroup(ctx, db, input.GroupID)
	if err != nil {
		return models.APITokenOut{}, err
	}

	_, err = db.ExecContext(ctx,
		"UPDATE api_tokens SET name = ?, description = ?, expires_at = ?, group_id = ?, updated_at = datetime('now') WHERE id = ?",
		trimSpace(input.Name), trimSpace(input.Description), expiresAt, groupID, id)
	if err != nil {
		return models.APITokenOut{}, apperr.Database(err)
	}

	return reloadToken(ctx, db, id)
}

func SetTokenEnabled(ctx context.Context, db *sql.DB, id int64, enabled bool) (models.APITokenOut, error) {
	_, err := db.ExecContext(ctx,
		"UPDATE api_tokens SET enabled = ?, updated_at = datetime('now') WHERE id = ?",
		boolToInt64(enabled), id)
	if err != nil {
		return models.APITokenOut{}, apperr.Database(err)
	}
	return reloadToken(ctx, db, id)
}

func reloadToken(ctx context.Context, db *sql.DB, id int64) (models.APITokenOut, error) {
	token, ok, err := GetToken(ctx, db, id)
	if err != nil {
		return models.APITokenOut{}, err
	}
	if !ok {
		return models.APITokenOut{}, apperr.NotFound("token not found")
	}
	return token, nil
}

func DeleteToken(ctx context.Context, db *sql.DB, id int64) (bool, error) {
	return execAffectsAny(ctx, db, "DELETE FROM api_tokens WHERE id = ?", id)
}

func equalOptionalString(left, right *string) bool {
	if left == nil || right == nil {
		return left == nil && right == nil
	}
	return *left == *right
}
